### PR TITLE
Add support for relationships

### PR DESCRIPTION
**Pull request by christoph-kluge**
Thanks for creating this package 👍

I'm adding this realtionship-support because I'm in need of this in a tenant-based stats-approach. The following discussion https://github.com/spatie/laravel-stats/discussions/5 was very useful for it. After some initial thoughts on how to achieve this I remembered how easy it is with `spatie/laravel-query-bullder`. I took this as an inspiration.

Feel free to follow up on the commit messages which may explain some changes.

### CODE REVIEW

File: src/Models/Stat.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+use Illuminate\Database\Eloquent\Model;
+
+class Stat extends Model
+{
+
+    public function events()
+    {
+        return $this->hasMany(StatsEvent::class);
+    }
+}
```
Comment (AlexVanderbist) [inline src/Models/Stat.php:14]: Is this model only used for tests?

File: database/migrations/create_stats_tables.php.stub [hunk 2]
```diff
@@ -8,10 +8,16 @@ class CreateStatsTables extends Migration
 {
     public function up()
     {
+        Schema::create('stats', function (Blueprint $table) {
+            $table->id();
+            $table->timestamps();
+        });
+
```
Comment (AlexVanderbist) [inline database/migrations/create_stats_tables.php.stub:15]: Is this migration only used for tests?

File: tests/StatsWriterTest.php [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+//
+//    /** @test */
+//    public function it_can_generate_and_array_of_periods()
+//    {
+//        $periods = StatsQuery::for(OrderStats::class)->start(now()->subYear())->end(now())->generatePeriods();
+//
+//        $this->assertCount(53, $periods);
+//
+//        $this->assertEquals([
+//            Carbon::parse('2018-12-31'),
```
Comment (AlexVanderbist) [inline tests/StatsWriterTest.php:203]: Are these old tests that still need fixing or can they be removed?

**Review by AlexVanderbist — COMMENTED** [review]
Hey, thanks for the great PR! Well documented and adds a lot of value 👌 I've added a couple questions in the PR review but apart from that, this looks good to merge

File: src/Models/Stat.php (hunk 1 above)
Comment (christoph-kluge) [reply src/Models/Stat.php:14]: Yes this is used only for tests. I didn't had an idea how to create a model (+ migration) only for tests.

File: database/migrations/create_stats_tables.php.stub (hunk 2 above)
Comment (christoph-kluge) [reply database/migrations/create_stats_tables.php.stub:15]: Yes this is used only for tests.

File: tests/StatsWriterTest.php (hunk 3 above)
Comment (christoph-kluge) [reply tests/StatsWriterTest.php:203]: Good spot! Saw that yesterday after re-reviewing my own stuff and almost forgot about it. 

I will update those. Perhaps I will move visibility to protected for `generatePeriods` as it's only used internally. I will add a set of new tests to cover this via `StatsWriter::get()`

File: src/Models/Stat.php (hunk 1 above)
Comment (christoph-kluge) [reply src/Models/Stat.php:14]: @AlexVanderbist should I do this differently? Are you aiming for a different approach here?

File: tests/StatsWriterTest.php (hunk 3 above)
Comment (AlexVanderbist) [reply tests/StatsWriterTest.php]: Awesome, thanks!

File: src/Models/Stat.php (hunk 1 above)
Comment (christoph-kluge) [reply src/Models/Stat.php]: That worked out like a breeze and I learned something new. Thank you! 🙏  I guess we're good to go now!

**Comment by christoph-kluge** [comment]
@AlexVanderbist thanks for reviewing 👍 In my opinion we're good to go. 

After looking at the existing code I saw that `StatsQuery->getPeriodTimestampFormat` is open to public but only used internally. Since we'll be increasing to `2.x` I would consider moving this to `protected`.

What do you think?

**Comment by AlexVanderbist** [comment]
Perfect, LGTM! I'll merge, make `getPeriodTimestampFormat` protected and tag `2.0.0`! Thanks for your great work on this :D

**Pull request merged**